### PR TITLE
Update the applications management UI to support setting scope permissions

### PR DESCRIPTION
**Pull request by Tsjerno**
Scope permissions are enforced. 
Updated the management UI is to allow defining the scopes an application is allowed to request.
See issue https://github.com/OrchardCMS/OrchardCore/issues/1981

### CODE REVIEW

**Comment by agriffard** [comment]
The Tests are in error.

**Comment by agriffard** [comment]
Remaining error on `UsersShouldBeAbleToCreateApplicationIfAllowed` test.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs [hunk 1]
```diff
@@ -113,7 +113,7 @@ public async Task<IActionResult> Create(string returnUrl = null)
                 _notifier.Warning(H["There are no registered services to provide roles."]);
             }
 
-            await foreach (var scope in _scopeManager.ListAsync())
+            await foreach (var scope in _scopeManager.ListAsync(null, null, default))
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs:116]: Curious: why this change was required?

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs [hunk 2]
```diff
@@ -54,7 +54,7 @@ public void Configure(OpenIddictServerOptions options)
                 return;
             }
 
-            options.IgnoreScopePermissions = true;
+            options.IgnoreScopePermissions = false;
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs:57]: `false` is the default value, so we can remove this line.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs [hunk 3]
```diff
@@ -287,6 +302,17 @@ public async Task<IActionResult> Edit(string id, string returnUrl = null)
                 _notifier.Warning(H["There are no registered services to provide roles."]);
             }
 
+            var scopes = await _applicationManager.GetPermissionsAsync(application);
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs:305]: `permissions` may be a better name here. Alternatively, you can also use the `_applicationManager.HasPermission(...)` method in the foreach body.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs [hunk 4]
```diff
@@ -483,13 +509,21 @@ await _applicationManager.GetIdAsync(application), StringComparison.Ordinal))
...
 
-            foreach (string selectedRole in (model.RoleEntries
+            foreach (var selectedRole in (model.RoleEntries
                 .Where(role => role.Selected)
                 .Select(role => role.Name)))
             {
                 descriptor.Roles.Add(selectedRole);
             }
 
+            descriptor.Permissions.RemoveWhere(scope => scope.StartsWith(OpenIddictConstants.Permissions.Prefixes.Scope));
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs:519]: `permission => permission`

**Review by kevinchalet — APPROVED** [review]
Looks great, thanks. Just 1 or 2 nits and we can merge it.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs (hunk 1 above)
Comment (Tsjerno) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs:116]: Otherwise the tests failed. I could not Mock a ListAsync(). If you have a suggestion for the unittest, then I can adjust the unittest. It looks kind a strange with this parameters

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs (hunk 2 above)
Comment (Tsjerno) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs:57]: Removed the line

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs (hunk 3 above)
Comment (Tsjerno) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs:305]: Will rename

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs (hunk 4 above)
Comment (Tsjerno) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs:519]: Renamed
